ouroboros-consensus/docs/Contributing.md

```diff
@@ -23,3 +23,131 @@ start.
+**NOTE**
+
+In order to explain this we have to take some detour on FFI. The following table
+shows some context about it:
+
+|                            | `safe` | `unsafe` |
+|----------------------------|--------|----------|
+| Can call into Haskell code | Yes    | No       |
+| Releases the capability    | Yes    | No       |
+| Shows up in time profiling | No     | Yes      |
+
+In particular, for `safe` calls, as the capability is released just before the
+foreign function is called, from the Haskell point of view, said task is not
+running. 
+
+Imagine the following code in Haskell:
+
+```haskell
+foreign import ccall unsafe "foo" foo :: IO ()
+foreign import ccall safe   "bar" foo :: IO ()
+
+baz = do
+  foo
+  bar
+```
+
+This will be compiled to Core (`-ddump-simpl`) as:
+
+```
+...
+ case {__pkg_ccall main State# RealWorld -> (# State# RealWorld #)}_d20O
+                        ds_X21f
+                 of
+...
+case {__pkg_ccall_GC main State# RealWorld -> (# State# RealWorld #)}_d20S
+             ipv4_X284
+      of
+...
+```
+
+And to C-- (`-ddump-cmm`) as:
+
+```
+...
+           call "ccall" arg hints:  []  result hints:  [] foo();
+...
+           (_u3jD::I64) = call "ccall" arg hints:  [PtrHint,]  result hints:  [PtrHint] suspendThread(BaseReg, 0);
+           call "ccall" arg hints:  []  result hints:  [] bar();
+           (_u3jE::I64) = call "ccall" arg hints:  [PtrHint]  result hints:  [PtrHint] resumeThread(_u3jD::I64);
+...
+```
+
+This shows how a `safe` call indeed releases the capability. Imagining a super
+simplified scenario with initially one OS thread, one Haskell task/thread and
+only one capability that performs the `baz` call:
+
+```
+= has the capability
+- not
+
+           v call to `foo`      v call to `bar`
+...===============================-------------------========= ...
+                              ^ return             ^ return
+```
+
+So in the above, the time during the `safe` call, the Haskell capability is
+`IDLE` thus not counting for the time spent. What happens if we have more
+Haskell tasks waiting to be executed (note this will require multiple OS
+threads)?
+
+```
+= has the capability
+- not
+
+           v call to `foo`      v call to `bar`
+...===============================----------------------========= ...
+                              ^ return             ^ return
+...-----------------------------========================-------- ...
+```
+
+Note now that the foreign call might finish before the capability is available
+again, and in this case it has to wait.
+
+--------------------------------------------------------------------------------
+
+Now, the reported time on the Haskell time profiling will be off due to these
+`safe` foreign calls not being accounted for. Apart from that, `unsafe` calls
+sometimes lose ticks when running with the default tick interval. Therefore it
+is advised to run with `-V0.01` when doing measurements on time.
```

Comment:
FTR: I opened https://gitlab.haskell.org/ghc/ghc/-/merge_requests/8186 to correct the incorrect default value mentioned [here](https://downloads.haskell.org/ghc/8.10.7/docs/html/users_guide/profiling.html#rts-flag--V%20%E2%9F%A8secs%E2%9F%A9) as discovered by Nick.